c_src/optics_htable/htable.c

```diff
@@ -0,0 +1,224 @@
+    return key;
+}
+
+struct Key *make_key_str(char *str)
+{
+    size_t len = strnlen(str, htable_key_max_len);
+    struct Key *key = malloc(sizeof(struct Key) + len);
+    assert(key != NULL);
+    key->n = len;
+    memcpy(key->bytes, str, len);
+    return key;
+}
+
+
+// -----------------------------------------------------------------------------
+// ops
+// -----------------------------------------------------------------------------
+
+struct htable_ret htable_get(struct htable *ht, const struct Key *key)
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (!bucket->key) continue;
+        if (bucket->key->n != key->n) continue;
+        if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+
+        return (struct htable_ret) { .ok = true, .value = bucket->value };
+    }
+
+    return (struct htable_ret) { .ok = false };
+}
+
+struct htable_ret htable_put(struct htable *ht, const struct Key *key, uint64_t value)
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    struct htable_bucket *empty = NULL;
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (bucket->key) {
+            if (bucket->key->n != key->n) continue;
+            if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+            return (struct htable_ret) { .ok = false, .value = bucket->value };
+        }
+
+        if (!empty) empty = bucket;
+    }
+
+    if (empty) {
+        ht->len++;
+        empty->key = key;
+        empty->value = value;
+        return (struct htable_ret) { .ok = true };
+    }
+
+    htable_resize(ht, ht->cap * 2);
+    return htable_put(ht, key, value);
+}
+
+struct htable_ret htable_del(struct htable *ht, const struct Key *key)
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (!bucket->key) continue;
+        if (bucket->key->n != key->n) continue;
+        if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+
+        ht->len--;
+        free((struct Key *) bucket->key);
+        bucket->key = NULL;
+        return (struct htable_ret) { .ok = true, .value = bucket->value };
+    }
+
+    return (struct htable_ret) { .ok = false };
+}
+
+
+struct htable_bucket * htable_next(
+        struct htable *ht, struct htable_bucket *bucket)
+{
+    if (!ht->table) return NULL;
+
+    size_t i = 0;
+    if (bucket) i = (bucket - ht->table) + 1;
+
+    for (; i < ht->cap; ++i) {
+        bucket = &ht->table[i];
+        if (bucket->key) return bucket;
+    }
+
+    return NULL;
+}
+
+void htable_itr_init(struct htable *ht, struct htable_iterator *itr)
```

Comment:
Here's an example of how I use `htable_next` (could be documented a little better): https://github.com/RAttab/optics/blob/master/test/htable_test.c#L119

In other words, I don't think the extra utilities are necessary as the interface already allows for pretty compact iteration but your millage may vary.

Reply:
True, it is more compact, but is it better interface-wise? I mean you're using a bucket pointer as an iterator, and you know it works because you know about the details of the htable_next function. But in general, a bucket is a bucket and its implementation could change, in which case, the bucket iterator won't work anymore.

Reply:
If you intend on hiding `struct bucket` from the interface (this is subjective so your call) then I would rework the interface to look something like this:

```
// If this is an implementation detail then don't make it public. 
// Same for `htable_next` which should not be in the header at all.
struct htable_bucket;

// Needs to be defined (not opaque) so it can be stack allocated
// I'm pretty sure these are the only fields that are actually needed but I might be wrong.
struct htable_it { struct htable *ht; struct htable_bucket *bucket; };

bool htable_it_next(struct htable_it *it); // returns false if there are no more elements
struct key * htable_it_key(const struct htable_it *it);
uint64_t htable_it_val(const struct htable_it *it);

struct htable_it htable_begin(struct htable *ht);
```
This version works best with a while loop. Adding an `htable_it_end` function and having next return `void` would make it work well with for-loops. Tweak at your discretion.

Reply:
What do you think about putting the key and value (pointers) in the iterator and remove htable_it_key and htable_it_val functions?

Reply:
I  feel that you're just back to having the old `struct htable_bucket` implementation but with an extra layer of indirection. Note that I usually tweak an interface as I'm using it to suite my needs so feel free to tweak as you see fit.

Note that my original version was used extensively and has worked quite well for me in the past :)